### PR TITLE
Add a CircleCI workflow for arm64

### LINKED ISSUE #332: Add arm64 CI
Possibly on CircleCI (https://circleci.com/docs/using-arm) or on ault (through the same setup as for #67).

### PR DESCRIPTION
**Pull request by aurianer**
Attempt to fix #332.
Open usage with M1 (untested so possibly broken).

### CODE REVIEW

**Comment by msimberg** [comment]
The build says "Resource class "arm.xlarge" is not a valid resource class. The default resource class will be used." Are you sure this is actually running on arm?

File: CMakeLists.txt [hunk 1]
```diff
@@ -1500,16 +1500,13 @@ else()
     OR "${__target_arch}" STREQUAL "armv5"
     OR "${__target_arch}" STREQUAL "armv6"
     OR "${__target_arch}" STREQUAL "armv7"
+    OR "${__target_arch}" STREQUAL "armv64"
```
Comment (msimberg) [inline CMakeLists.txt:1503]:
```suggestion
    OR "${__target_arch}" STREQUAL "arm64"
```

Comment (aurianer) [reply CMakeLists.txt:1503]: Thanks!

**Comment by aurianer** [comment]
> The build says "Resource class "arm.xlarge" is not a valid resource class. The default resource class will be used." Are you sure this is actually running on arm?

I added a print of the architecture so let's see.. in case it's arm it might be that this resource is not available in the free plan

File: .circleci/config.yml [hunk 2]
```diff
@@ -582,6 +582,55 @@ jobs:
           working_directory: /pika/build
           when: always
 
+  arm64_build:
+    <<: *docker_default
```
Comment (msimberg) [inline .circleci/config.yml:586]: You most likely don't want this. The image is not built for arm.

File: .circleci/config.yml [hunk 3]
```diff
@@ -582,6 +582,63 @@ jobs:
...
+    machine:
+      image: ubuntu-2004:current
+    resource_class: arm.large
+    steps:
+      - checkout
+      - run:
+          name: Building Arm64
+          command: |
+            # Check architecture = arm
+            uname -m
```
Comment (msimberg) [inline .circleci/config.yml:595]: This can probably be removed now?

File: .circleci/config.yml [hunk 4]
```diff
@@ -582,6 +582,63 @@ jobs:
...
+            # Install CMake
+            wget -q https://cmake.org/files/v3.22/cmake-3.22.6-linux-aarch64.tar.gz \
+              -O cmake.tar.gz && echo \
+              "79be85d3e76565faacd60695cee11d030f7e7dd8691476144fa25eb93dbd0397  cmake.tar.gz" \
+              | sha256sum --check --status && \
+              sudo tar --strip-components=1 -C /usr/local -xf cmake.tar.gz && \
+              rm cmake.tar.gz
+            mkdir build
+            git clone https://github.com/pika-org/pika.git
```
Comment (msimberg) [inline .circleci/config.yml:604]: This should be removed and use the `checkout` step instead.

File: .circleci/config.yml [hunk 5]
```diff
@@ -582,6 +582,63 @@ jobs:
...
+                -DPIKA_WITH_STACKTRACES_STATIC_SYMBOLS=On \
+                -DPIKA_WITH_STACKTRACES_DEMANGLE_SYMBOLS=Off \
+                -DPIKA_WITH_TESTS_DEBUG_LOG=On \
+                -DPIKA_WITH_TESTS_DEBUG_LOG_DESTINATION=/build/debug-log-arm.txt \
+                -DPIKA_WITH_SPINLOCK_DEADLOCK_DETECTION=On
+            ninja -j4 -k 0 pika
+      - persist_to_workspace:
+          root: /home/circleci/project
+          paths:
+            - ./build
```
Comment (msimberg) [inline .circleci/config.yml:640]: I don't think this is needed since we're not splitting this into multiple jobs (right now at least...).

File: .circleci/config.yml [hunk 6]
```diff
@@ -582,6 +582,63 @@ jobs:
...
+                -DPIKA_WITH_COMPILER_WARNINGS=On \
+                -DPIKA_WITH_COMPILER_WARNINGS_AS_ERRORS=On \
+                -DPIKA_WITH_DEPRECATION_WARNINGS=On \
+                -DPIKA_WITH_THREAD_LOCAL_STORAGE=On \
+                -DPIKA_WITH_STACKTRACES_STATIC_SYMBOLS=On \
+                -DPIKA_WITH_STACKTRACES_DEMANGLE_SYMBOLS=Off \
+                -DPIKA_WITH_TESTS_DEBUG_LOG=On \
+                -DPIKA_WITH_TESTS_DEBUG_LOG_DESTINATION=/build/debug-log-arm.txt \
+                -DPIKA_WITH_SPINLOCK_DEADLOCK_DETECTION=On
+            ninja -j4 -k 0 pika
```
Comment (msimberg) [inline .circleci/config.yml:636]: Could you please split this up into "Install dependencies", "Configure", "Build", and a new "Test" step? The "Test" step is optional to get working for this PR. I'd just like to see how good/bad the test runs actually are. If it works right away let's keep it in this PR. If there's _anything_ that looks off we delay it for a separate PR.

TL;DR. Splitting into multiple steps: required for this PR. Building and running tests: optional.

File: .circleci/config.yml [hunk 7]
```diff
@@ -654,3 +711,4 @@ workflows:
             - tests.regressions
             - tests.performance
             - tests.headers
+      - arm64_build
```
Comment (msimberg) [inline .circleci/config.yml:714]: I don't know if this works, but could this be another "top-level workflow" on the same level as `build-and-test`? Perhaps with the name `build-and-test-arm64` (or just `build-arm64` if we don't get the testing part yet in this PR).

**Review by msimberg — CHANGES_REQUESTED** [review]
Could you also add the arm64 step(s) to `bors.toml`?

Thanks for doing this, this is already very promising!

File: .circleci/config.yml (hunk 3 above)
Comment (msimberg) [reply .circleci/config.yml:595]: Ok, no worries. We can leave it.

File: .circleci/config.yml (hunk 4 above)
Comment (msimberg) [reply .circleci/config.yml:604]: What exactly do you mean by it doesn't work? The directory is not visible after cloning? If it doesn't work this will have to be changed to actually check out the branch that triggers the build.

File: .circleci/config.yml (hunk 5 above)
Comment (msimberg) [reply .circleci/config.yml:640]: We can add it again if and when we do run the tests in separate steps.

File: .circleci/config.yml [hunk 8]
```diff
@@ -654,3 +724,4 @@ workflows:
...
```
Comment (msimberg) [inline .circleci/config.yml:729]:
```suggestion
      - arm64_build:
          <<: *branch_filter
```

File: .circleci/config.yml [hunk 9]
```diff
@@ -582,6 +587,71 @@ jobs:
...
+                -DPIKA_WITH_STACKTRACES_STATIC_SYMBOLS=On \
+                -DPIKA_WITH_STACKTRACES_DEMANGLE_SYMBOLS=Off \
+                -DPIKA_WITH_TESTS_DEBUG_LOG=On \
+                -DPIKA_WITH_TESTS_DEBUG_LOG_DESTINATION=/build/debug-log-arm.txt \
+                -DPIKA_WITH_SPINLOCK_DEADLOCK_DETECTION=On
+          working_directory: /home/circleci/project/build
+      - run:
+          name: Build pika Arm64
+          command: |
+            ninja -j4 -k 0 pika
```
Comment (msimberg) [inline .circleci/config.yml:652]: Let's try if this can build without running out of memory:
```suggestion
            ninja -k 0 pika
```

**Review by msimberg — CHANGES_REQUESTED** [review]
Looks good. Now it's only the `bors.toml` change and `branch_filter` missing.

File: .circleci/config.yml [hunk 10]
```diff
@@ -582,6 +587,67 @@ jobs:
...
+            # Install pika dependencies
+            sudo apt install clang ninja-build libboost-regex-dev \
+              libboost-context-dev libhwloc-dev mpi-default-dev -y
+            mkdir build
+          working_directory: /home/circleci/project
+      - run:
+          name: Configure pika Arm64
+          command: |
+            cmake \
+                /home/circleci/project/pika \
```
[the remark this answers is not in the document]
Comment (msimberg) [reply .circleci/config.yml:617]: This wasn't supposed to here anymore... Please ignore.

**Review by msimberg — APPROVED** [review]
